### ISSUE TITLE
solc compiler bug related to character encoding of the source

### ISSUE THREAD

**Issue by pieterhartel** [ethereum/solidity#4614]
I have written a 4-line contract called foo.sol (below), which confuses the compiler. The first problem is that the error message points at the wrong text because the compiler does not know how many bytes a special character (the copyright symbol) needs. There are five occurrences of the copyright symbol and the arrow to the offending text is five positions off. The second problem is, that this bug also causes the AST generated by the compiler to be incorrect. This will impact tooling that builds on the AST.

--pieter

```
$ cat foo.sol 
pragma solidity ^0.4.18 ;
contract Foo {
/* ⒸⒸⒸⒸⒸ2017 */ constructor () public { true ; }
}
$ od -c foo.sol 
0000000    p   r   a   g   m   a       s   o   l   i   d   i   t   y    
0000020    ^   0   .   4   .   1   8       ;  \n   c   o   n   t   r   a
0000040    c   t       F   o   o       {  \n   /   *       Ⓒ  **  **   Ⓒ
0000060   **  **   Ⓒ  **  **   Ⓒ  **  **   Ⓒ  **  **   2   0   1   7    
0000100    *   /       c   o   n   s   t   r   u   c   t   o   r       (
0000120    )       p   u   b   l   i   c       {       t   r   u   e    
0000140    ;       }  \n   }  \n                                        
0000146
$ solc foo.sol 
foo.sol:3:51: Warning: Statement has no effect.
/* ⒸⒸⒸⒸⒸ2017 */ constructor () public { true ; }
                                                  ^--^
$ solc --version
solc, the solidity compiler commandline interface
Version: 0.4.24+commit.e67f0147.Darwin.appleclang
$ 
```

**Comment by chriseth** [comment]
Can you please be more specific about the impact on the AST?

Since the encoding of the input file is not specified ,this is hard to do, but I guess we could just always assume utf8.

**Comment by randomnetcat** [comment]
AST as generated by `0.4.25-develop.2018.8.3+commit.2c2d4c47.Linux.g++`:
```
======= <stdin> =======
PragmaDirective
   Gas costs: 0
   Source: "pragma solidity ^0.4.18 ;"
ContractDefinition "Foo"
   Source: "contract Foo {\n/* \u24b8\u24b8\u24b8\u24b8\u24b82017 */ constructor () public { true ; }\n}"
  FunctionDefinition "" - public
     Source: "constructor () public { true ; }"
    ParameterList
       Gas costs: 0
       Source: "()"
    ParameterList
       Gas costs: 0
       Source: ""
    Block
       Source: "{ true ; }"
      ExpressionStatement
         Gas costs: 0
         Source: "true"
        Literal, token: true value: true
           Type: bool
           Source: "true"
```

**Comment by pieterhartel** [comment]
The JSON style AST contains the position in the source text of the lexical items. These positions are also off because of the UTF-8 issue. (See example AST node below).

I am using the JSON style AST for program transformations on solidity contracts. This works fine if the source is in ASCII. If it is not, I have to convert it as a work around first (with iconv) loosing the special symbols. Since these usually occur in the comments, this does not appear to be a big problem (yet).

In any case a fix would be much appreciated.

--pieter

{                                                                               
        "attributes" :                                                                  
        {                                                                               
                "argumentTypes" : null,                                                         
                "hexvalue" : "74727565",                                                        
                "isConstant" : false,                                                           
                "isLValue" : false,                                                             
                "isPure" : true,                                                                
                "lValueRequested" : false,                                                      
                "subdenomination" : null,                                                       
                "token" : "bool",                                                               
                "type" : "bool",                                                                
                "value" : "true"                                                                
        },                                                                              
        "id" : 4,                                                                       
        "name" : "Literal",                                                             
        "src" : "91:4:0"                                                                
}

**Comment by pieterhartel** [comment]
I have added a comment to the thread.

—pieter

**Comment by axic** [comment]
> Since the encoding of the input file is not specified ,this is hard to do, but I guess we could just always assume utf8.

We do require UTF-8 at least in string literals.

I guess what could be improved is offering two source locations:
- actual byte offset
- character offset (parsed as UTF-8)

For the second, we would need to parse everything, including comments, as UTF-8 however I believe comments are not properly parsed at the moment.

**Comment by chriseth** [comment]
I think it is fine to only work with byte offsets. What we could improve is the calculation of the `^--^` output - and note that tabs are a similar complication we should take into account, but perhaps first check what other compilers do.

**Comment by chriseth** [comment]
Fixed (well enough) by #9012
